Add CICD spans and resources (#2013)
open-telemetry/semantic-conventions: docs/resource/cicd.md

**Status:** ![Development](https://img.shields.io/badge/-development-blue)

**type:** `cicd.pipeline`

**Description:** A pipeline is a series of automated steps that helps software teams deliver code.

| Attribute  | Type | Description  | Examples  | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) | Stability |
|---|---|---|---|---|---|
| [`cicd.pipeline.name`](/docs/attributes-registry/cicd.md) | string | The human readable name of the pipeline within a CI/CD system. | `Build and Test`; `Lint`; `Deploy Go Project`; `deploy_to_environment` | `Recommended` | ![Development](https://img.shields.io/badge/-development-blue) |

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->
<!-- END AUTOGENERATED TEXT -->
<!-- endsemconv -->

### CICD pipeline run resource attributes

<!-- semconv resource.cicd.pipelinerun -->
<!-- NOTE: THIS TEXT IS AUTOGENERATED. DO NOT EDIT BY HAND. -->
<!-- see templates/registry/markdown/snippet.md.j2 -->
<!-- prettier-ignore-start -->
<!-- markdownlint-capture -->
<!-- markdownlint-disable -->

**Status:** ![Development](https://img.shields.io/badge/-development-blue)

> **type:** `cicd.pipelinerun`

Reviewer: ```suggestion
**type:** `cicd.pipeline.run`
```

Author: Renamed in 77cca2d9657f7c4b02bfafdc6c444be5aafd7a30

<!-- markdownlint-capture -->
<!-- markdownlint-disable -->

**Status:** ![Development](https://img.shields.io/badge/-development-blue)

**type:** `vcs.ref`

**Description:** A reference to a specific version in the Version Control System.

| Attribute  | Type | Description  | Examples  | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) | Stability |
|---|---|---|---|---|---|
| [`vcs.ref.head.name`](/docs/attributes-registry/vcs.md) | string | The name of the [reference](https://git-scm.com/docs/gitglossary#def_ref) such as **branch** or **tag** in the repository. [1] | `my-feature-branch`; `tag-1-test` | `Recommended` | ![Development](https://img.shields.io/badge/-development-blue) |
| [`vcs.ref.head.revision`](/docs/attributes-registry/vcs.md) | string | The revision, literally [revised version](https://www.merriam-webster.com/dictionary/revision), The revision most often refers to a commit object in Git, or a revision number in SVN. [2] | `9d59409acf479dfa0df1aa568182e43e43df8bbe28d60fcf2bc52e30068802cc`; `main`; `123`; `HEAD` | `Recommended` | ![Development](https://img.shields.io/badge/-development-blue) |

> | [`vcs.ref.type`](/docs/attributes-registry/vcs.md) | string | The type of the [reference](https://git-scm.com/docs/gitglossary#def_ref) in the repository. | `branch`; `tag` | `Recommended` | ![Development](https://img.shields.io/badge/-development-blue) |

Reviewer: A commit sha, Could be a type?

Author: The `vcs.ref.type` refers to the type of `vcs.ref.head.name` whether it refers to a branch or a tag.
The commit sha is stored in `vcs.ref.head.revision`.

Do you mean that `sha` could be an additional value for `vcs.ref.type`?

Reviewer: Yes, it the case `vcs.ref.head.revision` is a `sha` How we inform `vcs.ref.type`?  maybe `sha` or `id` or something along these lines

Reviewer 2: The `vcs.ref.head.revision` should always be set to the most precise revision which in most cases is the sha.

Since `vcs.ref.type` refers to the type of `vcs.ref.head.name` I would not add a value for sha unless `vcs.ref.head.name` itself contains a sha.
If `vcs.ref.head.name` is not set then I would leave `vcs.ref.type` unset as well even if `vcs.ref.head.revision` contains a sha.

<!-- semconv resource.cicd.pipeline -->
<!-- NOTE: THIS TEXT IS AUTOGENERATED. DO NOT EDIT BY HAND. -->
<!-- see templates/registry/markdown/snippet.md.j2 -->
<!-- prettier-ignore-start -->
<!-- markdownlint-capture -->
<!-- markdownlint-disable -->

**Status:** ![Development](https://img.shields.io/badge/-development-blue)

**type:** `cicd.pipeline`

**Description:** A pipeline is a series of automated steps that helps software teams deliver code.

| Attribute  | Type | Description  | Examples  | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) | Stability |
|---|---|---|---|---|---|
| [`cicd.pipeline.name`](/docs/attributes-registry/cicd.md) | string | The human readable name of the pipeline within a CI/CD system. | `Build and Test`; `Lint`; `Deploy Go Project`; `deploy_to_environment` | `Recommended` | ![Development](https://img.shields.io/badge/-development-blue) |

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->
<!-- END AUTOGENERATED TEXT -->
<!-- endsemconv -->

### CICD pipeline run

To efficiently set the following attributes for all spans of a pipeline run they should be defined as the following resource.

> Since this resource is unique for each pipeline run, this means instantiating separate [TracerProviders](https://opentelemetry.io/docs/specs/otel/trace/api/#tracerprovider) for each pipeline run in order to be able to configure the resources of any spans emitted through that tracer.

Author: Related to https://github.com/open-telemetry/semantic-conventions/issues/1111 https://github.com/open-telemetry/semantic-conventions/issues/1184#issuecomment-2773594630, we should add a warning here that if the `CICD pipeline run` resource is used for metrics, then it might cause issues with high-cardinality.

Author: I added a warning in 7f6b5b38b0620ea6ef75cfe24bffa4aa2b95cab7.

<!--- Hugo front matter used to generate the website version of this page:
linkTitle: CICD
--->

# CICD

<!-- toc -->

- [Continuous Integration and Continuous Deployment (CICD)](#continuous-integration-and-continuous-deployment-cicd)
  - [CICD pipeline](#cicd-pipeline)
  - [CICD pipeline run](#cicd-pipeline-run)
  - [CICD worker](#cicd-worker)
- [Version Control System (VCS)](#version-control-system-vcs)
  - [VCS repository](#vcs-repository)
  - [VCS reference](#vcs-reference)

<!-- tocstop -->

See also:

- The [Trace semantic conventions for CICD](/docs/cicd/cicd-spans.md)

## Continuous Integration and Continuous Deployment (CICD)

> CICD aims to increase early defect discovery, increase productivity, and provide faster release cycles through automating builds, tests and deployment of software.

Author: Not really, I will remove it...

> CICD aims to increase early defect discovery, increase productivity, and provide faster release cycles through automating builds, tests and deployment of software.

Author: Removed in 29f122accda138bed61d214e2333efbf2e82e6bf.

**type:** `cicd.pipeline`

**Description:** A pipeline is a series of automated steps that helps software teams deliver code.

| Attribute  | Type | Description  | Examples  | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) | Stability |
|---|---|---|---|---|---|
| [`cicd.pipeline.name`](/docs/attributes-registry/cicd.md) | string | The human readable name of the pipeline within a CI/CD system. | `Build and Test`; `Lint`; `Deploy Go Project`; `deploy_to_environment` | `Recommended` | ![Development](https://img.shields.io/badge/-development-blue) |

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->
<!-- END AUTOGENERATED TEXT -->
<!-- endsemconv -->

### CICD pipeline run

To efficiently set the following attributes for all spans of a pipeline run they should be defined as the following resource.

Since this resource is unique for each pipeline run, this means instantiating separate [TracerProviders](https://opentelemetry.io/docs/specs/otel/trace/api/#tracerprovider) for each pipeline run in order to be able to configure the resources of any spans emitted through that tracer.

> Using the CICD pipeline run resource with metrics inherently causes high cardinality and might cause issues with some metric storage backends. Thus using CICD pipeline run resource for metrics MUST be opt-in.

Reviewer 3: ```suggestion
Using the CICD pipeline run resource with metrics inherently causes high cardinality and may increase costs with some metric storage backends. Thus, using CICD pipeline run resource for metrics MUST be opt-in.
```

Two things:

1) We probably shouldn't have the cicd.pipeline.name attribute on every metric. 
2) We should provide guidance around attaching the metrics directly to the spans?

Author: I will update the PR with this suggested change...

> We probably shouldn't have the cicd.pipeline.name attribute on every metric

This PR does not add guidance which resource to attach to which signal.
It only says for spans "Any resources of the CICD and VCS resource conventions that apply SHOULD be used."

The cicd pipeline names may also cause cardinality issues if less of a problem than for pipeline runs.
Should I also add a similar warning for the pipeline resource "Using the CICD pipeline resource with metrics may cause issues with high cardinality..."?

> We should provide guidance around attaching the metrics directly to the spans?

What do you mean with attaching the metrics to the span?

| Attribute  | Type | Description  | Examples  | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) | Stability |
|---|---|---|---|---|---|
| [`cicd.worker.id`](/docs/attributes-registry/cicd.md) | string | The unique identifier of a worker within a CICD system. | `abc123`; `10.0.0.1`; `controller` | `Recommended` | ![Development](https://img.shields.io/badge/-development-blue) |
| [`cicd.worker.name`](/docs/attributes-registry/cicd.md) | string | The name of a worker within a CICD system. | `agent-abc123`; `controller` | `Recommended` | ![Development](https://img.shields.io/badge/-development-blue) |
| [`cicd.worker.url.full`](/docs/attributes-registry/cicd.md) | string | The [URL](https://wikipedia.org/wiki/URL) of the worker, providing the complete address in order to locate and identify the worker. | `https://cicd.example.org/worker/abc123` | `Opt-In` | ![Development](https://img.shields.io/badge/-development-blue) |

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->
<!-- END AUTOGENERATED TEXT -->
<!-- endsemconv -->

## Version Control System (VCS)

A Version Control System tracks changes to source code.

> ### VCS repository

Author: The discussion https://github.com/open-telemetry/semantic-conventions/pull/2013#discussion_r2016086021 about needing to instantiate several TracerProviders also applies to the VCS resource especially for vcs.ref

I would prefer to resolve this discussion in this PR for both cicd and vcs resources.

What benefit is there in splitting off the vcs part? Which parts of this PR relating to VCS do you propose to split off in its own PR?
